x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SLMGetExpiredSnapshotsAction.java

```diff
@@ -0,0 +1,304 @@
+
+                                // Get repository data
+                                .newForked(repository::getRepositoryData)
+
+                                // Collect snapshot details by policy, and get any missing details by reading SnapshotInfo
+                                .<SnapshotDetailsByPolicy>andThen(
+                                    retentionExecutor,
+                                    threadContext,
+                                    (l, repositoryData) -> getSnapshotDetailsByPolicy(repository, repositoryData, l)
+                                )
+
+                                // Compute snapshots to delete for each (relevant) policy
+                                .<Void>andThen(
+                                    retentionExecutor,
+                                    threadContext,
+                                    (l, snapshotDetailsByPolicy) -> ActionListener.completeWith(l, () -> {
+                                        resultsBuilder.addResult(
+                                            repositoryName,
+                                            getSnapshotsToDelete(repositoryName, request.policies(), snapshotDetailsByPolicy)
+                                        );
+                                        return null;
+                                    })
+                                )
+
+                                // And notify this repository's listener on completion
+                                .addListener(perRepositoryListener.delegateResponse((l, e) -> {
+                                    logger.debug(Strings.format("[%s]: could not compute expired snapshots", repositoryName), e);
+                                    l.onResponse(null);
+                                }))
+                        )
+                    );
+                }
+            }
+        }
+    }
+
+    private static class SnapshotDetailsByPolicy {
+        private final Map<String, Map<SnapshotId, RepositoryData.SnapshotDetails>> snapshotsByPolicy = new HashMap<>();
+
+        synchronized void add(SnapshotId snapshotId, RepositoryData.SnapshotDetails snapshotDetails) {
+            assert RETAINABLE_STATES.contains(snapshotDetails.getSnapshotState());
+            final var slmPolicy = snapshotDetails.getSlmPolicy();
+            if (Strings.hasLength(slmPolicy)) {
+                snapshotsByPolicy.computeIfAbsent(slmPolicy, ignored -> new HashMap<>()).put(snapshotId, snapshotDetails);
+            }
+        }
+
+        <T> Stream<T> flatMap(BiFunction<String, Map<SnapshotId, RepositoryData.SnapshotDetails>, Stream<T>> fn) {
+            // no need for synchronized here, building the map entirely happens-before reading it
+            return snapshotsByPolicy.entrySet().stream().flatMap(entry -> fn.apply(entry.getKey(), entry.getValue()));
+        }
+    }
+
+    private static void getSnapshotDetailsByPolicy(
```

Comment:
Same as below, perhaps a personal preference, but should we add a bit more targeted unit testing to this function?

Reply:
Likewise, see eef1e63b42609a8c5612d665a03b2520c091f774.